### PR TITLE
Fix order by positional arg in case unneeded columns pruning

### LINKED ISSUE #43964: Removing unneeded columns from CTE subquery confuses interpretation of positional argument(s)
**Description**
Selecting a subset of columns contained in a CTE subquery throws a 'Positional argument out of bounds' exception.  Can be reproduced on v22.11.

**Reproduction Steps**
Executing this query against the 'UK Property Price Paid' sample dataset should trigger the error:
```
with top_counties as (
    select county,
        count(*) as transactions
    from
        pp_complete
    where
        toStartOfMonth(parseDateTimeBestEffort(time)) = '2019-09-01'
    group by
        1
    order by
        2 desc
    limit
        5
)
select county from top_counties
```
**Error message:**
```
"Code: 43. DB::Exception: Positional argument out of bounds: 2 (exprected in range [1, 1]. (ILLEGAL_TYPE_OF_ARGUMENT) (version 22.11.1.19280 (official build))\n"
```

**Expected result:**
| county             |
|--------------------|
| GREATER LONDON     |
| GREATER MANCHESTER |
| WEST YORKSHIRE     |
| WEST MIDLANDS      |
| ESSEX              |

### PR DESCRIPTION
**Pull request by kssenii**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix order by positional arg in case unneeded columns pruning. Closes https://github.com/ClickHouse/ClickHouse/issues/43964.

### CODE REVIEW

File: src/Interpreters/TreeRewriter.cpp [hunk 1]
```diff
@@ -325,6 +326,39 @@ struct ExistsExpressionData
...
+            if (select_query.groupBy())
+            {
+                for (auto & expr : select_query.groupBy()->children)
+                    replaceForPositionalArguments(expr, &select_query, ASTSelectQuery::Expression::GROUP_BY);
+            }
+            if (select_query.orderBy())
+            {
+                for (auto & expr : select_query.orderBy()->children)
+                {
+                    auto elem = assert_cast<ASTOrderByElement>(*expr).children.at(0);
```
Comment (kssenii) [inline src/Interpreters/TreeRewriter.cpp:347]: The fix is actually only in this line (extracting child of ASTOrderByElement). It was not necessary to add a visitor - it works in subquery without it. But it is now more consistent with the other code.

File: src/Interpreters/TreeRewriter.cpp [hunk 2]
```diff
@@ -325,6 +326,39 @@ struct ExistsExpressionData
...
+            if (select_query.groupBy())
+            {
+                for (auto & expr : select_query.groupBy()->children)
+                    replaceForPositionalArguments(expr, &select_query, ASTSelectQuery::Expression::GROUP_BY);
+            }
+            if (select_query.orderBy())
+            {
+                for (auto & expr : select_query.orderBy()->children)
+                {
+                    for (auto & elem : assert_cast<ASTOrderByElement &>(*expr).children)
```
Comment (kssenii) [inline src/Interpreters/TreeRewriter.cpp:347]: The fix is actually only in this line (extracting child of ASTOrderByElement). It was not necessary to add a visitor - it works in subquery without it. But it is now more consistent with the other code.

File: src/Interpreters/TreeRewriter.cpp [hunk 3]
```diff
@@ -325,6 +326,39 @@ struct ExistsExpressionData
...
 using ExistsExpressionVisitor = InDepthNodeVisitor<OneTypeMatcher<ExistsExpressionData>, false>;
 
+struct ReplacePositionalArgumentsData
+{
+    using TypeToVisit = ASTSelectQuery;
+    ContextPtr context;
+
+    void visit(ASTSelectQuery & select_query, ASTPtr &) const
+    {
+        if (context->getSettingsRef().enable_positional_arguments)
```
Comment (vdimir) [inline src/Interpreters/TreeRewriter.cpp:336]: We can check the setting in `TreeRewrtier::normalize` and do not apply `ReplacePositionalArgumentsVisitor` at all if `enable_positional_arguments = false` to improve readability a bit.

**Review by vdimir — APPROVED** [review]

**Comment by kssenii** [comment]
> Stress test (ubsan) — Fatal message in clickhouse-server.log

https://github.com/ClickHouse/ClickHouse/issues/44037

> Stress test (msan) — Killed by signal (in clickhouse-server.log)

<Fatal> : Logical error: 'Cannot commit shared transaction'.

**Pull request merged**